numpy/lib/_npyio_impl.py

```diff
@@ -1559,84 +1559,101 @@ def first_write(self, v):
-                format = delimiter.join(fmt)
-            elif iscomplex_X and n_fmt_chars != (2 * ncol):
-                raise error
-            elif ((not iscomplex_X) and n_fmt_chars != ncol):
-                raise error
+
+        if fstring_fmt:
+            if isinstance(fstring_fmt, str):
+                if "{" in fstring_fmt and "}" in fstring_fmt:
+                    if fstring_fmt.count("{") > 1 or fstring_fmt.count("}") > 1:
+                        raise ValueError(
+                            "Multi-column f-strings are not supported for fstring_fmt. "
+                            "Please provide a single f-string for formatting."
+                        )
+                fstring_fmt = [fstring_fmt] * ncol
+            elif isinstance(fstring_fmt, (list, tuple)):
+                raise ValueError(
+                    "Multi-column fstring_fmt is not supported. "
+                    "Provide a single string for fstring_fmt."
+                )
             else:
-                format = fmt
+                raise ValueError("fstring_fmt must be a string.")
+            format_row = delimiter.join(fstring_fmt)
         else:
-            raise ValueError('invalid fmt: %r' % (fmt,))
+            if isinstance(fmt, (list, tuple)):
+                if len(fmt) != ncol:
+                    raise ValueError(f"fmt has wrong shape: {fmt}")
+                format = delimiter.join(fmt)
+            elif isinstance(fmt, str):
+                n_fmt_chars = fmt.count('%')
+                if n_fmt_chars == 1:
+                    fmt = [fmt] * ncol
+                    format = delimiter.join(fmt)
+                elif iscomplex_X and n_fmt_chars != (2 * ncol):
+                    raise ValueError(f"fmt has wrong number of %% formats: {fmt}")
+                elif not iscomplex_X and n_fmt_chars != ncol:
+                    raise ValueError(f"fmt has wrong number of %% formats: {fmt}")
+                else:
+                    format = fmt
+            else:
+                raise ValueError(f"Invalid fmt: {fmt!r}")
 
-        if len(header) > 0:
-            header = header.replace('\n', '\n' + comments)
+        # Write header
+        if header:
+            header = header.replace("\n", "\n" + comments)
             fh.write(comments + header + newline)
-        if iscomplex_X:
+
+        # Write data rows
+        if fstring_fmt:
+            for row in X:
+                if iscomplex_X:
+                    row = [(item.real, item.imag) for item in row]
+                    row = [item for sublist in row for item in sublist]
+                row_formatted = format_row.format(*row)
+                fh.write(row_formatted + newline)
+        elif iscomplex_X:
             for row in X:
-                row2 = []
-                for number in row:
-                    row2.append(number.real)
-                    row2.append(number.imag)
+                row2 = [number.real if i % 2 == 0 else number.imag for i, number in
```

Comment:
Please revert these changes. They are not related to the fstring_fmt and (IMO) make the code less readable.